Fix - Error in redirects, connected to Axios?

Import trace for requested module:
./node_modules/follow-redirects/debug.js
./node_modules/follow-redirects/index.js
./node_modules/axios/lib/adapters/http.js
./node_modules/axios/lib/defaults.js
./node_modules/axios/lib/axios.js
./node_modules/axios/index.js
./app/context/spotify-context.js
 ⚠ ./node_modules/follow-redirects/debug.js
Module not found: Can't resolve 'debug' in '/Users/sabrinsettle/Desktop/Randofy/node_modules/follow-redirects'
Did you mean './debug'?
Requests that should resolve in the current directory need to start with './'.
Requests that start with a name are treated as module requests and resolve within module directories (node_modules, /Users/sabrinsettle/Desktop/Randofy).
If changing the source code is not an option there is also a resolve options called 'preferRelative' which tries to resolve these kind of requests in the current directory too.

--- a/app/_components/layout/header/AuthButton/AuthButton.jsx
+++ b/app/_components/layout/header/AuthButton/AuthButton.jsx
@@ -2,7 +2,7 @@ import { useSpotifyContext } from "../../../../context/spotify-context";
 
 export default function AuthButton() {
   const { spotifyClient } = useSpotifyContext();
-  const { spotifyUser } = spotifyClient;
+  const { spotifyUser, isLoading } = spotifyClient;
   // console.log(spotifyUser);
 
   // const buttonCn =
@@ -11,20 +11,31 @@ export default function AuthButton() {
   const buttonCn =
     "font-body bg-gray-100 text-body-md text-gray-700 px-2 py-1 rounded-sm border border-transparent hover:border-gray-400 transition-all duration-400 ease-in-out";
 
-  if (spotifyUser) {
-    return (
-      <button
-        onClick={() => spotifyClient.logoutRequest()}
-        className={buttonCn}
-      >
-        Sign Out
-      </button>
-    );
-  }
+  const handleClickAction = () => {
+    if (spotifyUser) {
+      spotifyClient.logoutRequest();
+    } else {
+      spotifyClient.loginRequest();
+    }
+  };
+
+  // if (spotifyUser) {
+  //   return (
+  //     <button onClick={() => handleClickAction()} className={buttonCn}>
+  //       Sign Out
+  //     </button>
+  //   );
+  // }
+
+  const btnText = spotifyUser ? "Sign Out" : "Connect with Spotify";
 
   return (
-    <button onClick={() => spotifyClient.loginRequest()} className={buttonCn}>
-      Connect with Spotify
+    <button
+      onClick={() => handleClickAction()}
+      className={buttonCn}
+      disabled={isLoading}
+    >
+      {isLoading ? "Loading..." : btnText}
     </button>
   );
 }

--- a/app/context/spotify-context.js
+++ b/app/context/spotify-context.js
@@ -79,15 +79,18 @@ export function SpotifyClientProvider({ children }) {
     // gets the user https://api.spotify.com/v1/me
     // https://developer.spotify.com/documentation/web-api/reference/users-profile/get-current-users-profile/
     // needs the access_token and token_type in the request
-    const res = await fetch(`/api/token?code=${_code}`, {
-      method: "GET",
-      headers: {
-        "Content-Type": "application/json",
-      },
-    });
-    if (!res.ok) {
-      throw new Error("Failed to fetch token");
-    } else {
+    try {
+      const res = await fetch(`/api/token?code=${_code}`, {
+        method: "GET",
+        headers: {
+          "Content-Type": "application/json",
+        },
+      });
+
+      if (!res.ok) {
+        throw new Error("Failed to fetch token");
+      }
+
       const data = await res.json();
       const seconds = data.expires_in;
       data.created_at = new Date();
@@ -95,12 +98,19 @@ export function SpotifyClientProvider({ children }) {
 
       localStorage.setItem("auth", JSON.stringify(data));
       setAuth(data);
-      getSpotifyUser(data);
+
+      await getSpotifyUser(data);
+      // console.log("User data fetched");
+    } catch (error) {
+      console.error("Token call error:", error);
+      setError(error.message);
+    } finally {
+      setIsLoading(false); // Always set loading false
     }
-    // needs error handling;
   };
 
   const getSpotifyUser = async (authData = auth) => {
+    console.log("Fetching user data");
     if (!authData) return;
 
     const res = await fetch("https://api.spotify.com/v1/me", {
@@ -316,6 +326,7 @@ export function SpotifyClientProvider({ children }) {
   };
 
   const loginRequest = async () => {
+    setIsLoading(true);
     window.location.href = "/api/login";
   };
 
@@ -339,6 +350,12 @@ export function SpotifyClientProvider({ children }) {
       onSuccessCode(_code);
     }
   }, [_code]);
+
+  useEffect(() => {
+    if (spotifyUser !== null) {
+      setIsLoading(false);
+    }
+  }, [spotifyUser]);
 
   const spotifyClient = {
     // State

--- a/app/context/spotify-context.js
+++ b/app/context/spotify-context.js
@@ -22,20 +22,6 @@ export function SpotifyClientProvider({ children }) {
   const { showToast } = useToast();
   const isMobile = useIsMobile();
 
-  const valueStrings = {
-    popularity: ["Unknown", "Kinda Known", "Known", "Famous"],
-    acoustics: [
-      "All Electric",
-      "Mostly Electric",
-      "Some Acoustic",
-      "All Acoustic",
-    ],
-    energy: ["Super Chill", "Kinda Chill", "Kinda Hype", "Super Hype"],
-    vocals: ["No Vocals", "Some Vocals", "Lots of Vocals", "All Vocals"],
-    danceability: ["No Groove", "Almost a Bop", "Bop", "Dance Party"],
-    mood: ["Real Low", "Kinda Low", "Kinda High", "Real High"],
-  };
-
   useEffect(() => {
     const auth = JSON.parse(localStorage.getItem("auth"));
     const spotifyUser = JSON.parse(localStorage.getItem("spotifyUser"));
@@ -66,7 +52,7 @@ export function SpotifyClientProvider({ children }) {
 
   const checkTokenTime = async () => {
     // checks if auth is present and if a new time is greater than the time for auth
-    if (auth && auth.expires_at && new Date() > auth.expires_at) {
+    if (auth && auth.expires_at && new Date() > new Date(auth.expires_at)) {
       const params = new URLSearchParams({
         refresh_token: auth.refresh_token,
       });
@@ -105,30 +91,29 @@ export function SpotifyClientProvider({ children }) {
       const data = await res.json();
       const seconds = data.expires_in;
       data.created_at = new Date();
-      data.expires_at = new Date().setSeconds(
-        data.created_at.getSeconds() + seconds,
-      );
+      data.expires_at = new Date(data.created_at.getTime() + seconds * 1000);
+
       localStorage.setItem("auth", JSON.stringify(data));
       setAuth(data);
-      getSpotifyUser();
+      getSpotifyUser(data);
     }
     // needs error handling;
   };
 
-  const getSpotifyUser = async () => {
-    if (!auth) return;
+  const getSpotifyUser = async (authData = auth) => {
+    if (!authData) return;
 
     const res = await fetch("https://api.spotify.com/v1/me", {
       headers: {
-        Authorization: `Bearer ${auth.access_token}`,
+        Authorization: `Bearer ${authData.access_token}`,
       },
     });
     const data = await res.json();
     // console.log(data);
     setSpotifyUser(data);
     localStorage.setItem("spotifyUser", JSON.stringify(data));
-    window.history.pushState(
-      "",
+    window.history.replaceState(
+      {},
       "",
       window.location.host.includes("localhost")
         ? "http://" + window.location.host
@@ -330,27 +315,6 @@ export function SpotifyClientProvider({ children }) {
     }
   };
 
-  // const updateSongHistory = (songs) => {
-  //   // localStorage.removeItem("history");
-  //   // check if local storage can be reached else return
-  //   const songHistory = JSON.parse(localStorage.getItem("history"));
-  //   const date = new Date();
-  //   const dateKey = date.toLocaleDateString();
-
-  //   if (!songHistory) {
-  //     localStorage.setItem("history", JSON.stringify({ [dateKey]: songs }));
-  //   } else {
-  //     if (!songHistory[dateKey]) {
-  //       songHistory[dateKey] = [];
-  //     }
-  //     const songList = [...songHistory[dateKey], ...songs];
-  //     songHistory[dateKey] = songList;
-  //     localStorage.setItem("history", JSON.stringify(songHistory));
-
-  //     setGenerationHistory(songHistory);
-  //   }
-  // };
-
   const loginRequest = async () => {
     window.location.href = "/api/login";
   };
@@ -370,70 +334,11 @@ export function SpotifyClientProvider({ children }) {
     console.log("Logged out.");
   };
 
-  React.useMemo(async () => {
+  useEffect(() => {
     if (_code) {
-      await onSuccessCode(_code);
+      onSuccessCode(_code);
     }
   }, [_code]);
-
-  // const getSongs = async () => {
-  //   setIsLoading(true);
-
-  //   const params = new URLSearchParams();
-
-  //   // const [songDetails, setSongDetails] = useState({
-  //   //   popularity: { min: 0, max: 100 },
-  //   //   acoustics: { min: 0.0, max: 1.0 },
-  //   //   energy: { min: 0.0, max: 1.0 },
-  //   //   vocals: { min: 0.0, max: 1.0 },
-  //   //   danceability: { min: 0.0, max: 1.0 },
-  //   //   mood: { min: 0.0, max: 1.0 },
-  //   // });
-
-  //   params.set("limit", songLimit);
-
-  //   params.set("min_popularity", Math.ceil(songDetails.popularity.min * 100));
-  //   params.set("max_popularity", Math.ceil(songDetails.popularity.max * 100));
-
-  //   params.set("min_energy", songDetails.energy.min);
-  //   params.set("max_energy", songDetails.energy.max);
-
-  //   params.set("min_danceability", songDetails.danceability.min);
-  //   params.set("max_danceability", songDetails.danceability.max);
-
-  //   params.set("min_acousticness", songDetails.acoustics.min);
-  //   params.set("max_acousticness", songDetails.acoustics.max);
-
-  //   params.set("min_speechiness", songDetails.vocals.min);
-  //   params.set("max_speechiness", songDetails.vocals.max);
-
-  //   params.set("min_valence", songDetails.mood.min);
-  //   params.set("max_valence", songDetails.mood.max);
-
-  //   if (!genres.size === 0) {
-  //     const fiveGenres = genres.size > 5 ? genres.slice(0, 5) : genres;
-  //     params.set("genres", Array.from(fiveGenres));
-  //   }
-
-  //   const res = await fetch("/api/random?" + params.toString());
-
-  //   if (!res.ok) {
-  //     setError(await res.json());
-  //     setIsLoading(false);
-  //   } else {
-  //     const data = await res.json();
-  //     // console.log(data);
-
-  //     let song = data.recommendedTracks[0];
-  //     // console.log(song);
-  //     setCurrentSongs(data.recommendedTracks);
-  //     setSelectedSong({ index: 0, song });
-
-  //     updateSongHistory(data.recommendedTracks);
-  //     setIsLoading(false);
-  //     return data;
-  //   }
-  // };
 
   const spotifyClient = {
     // State

